### PR TITLE
Port close-on-esc mixin

### PR DESCRIPTION
**Pull request by spen**
### Summary

This PR is a slightly spruced up & more complete version of #10268 (I've opened a new PR on an internal branch vs a fork).

You'll find some details and discussion over at #10268, and more up to date details below.

### Motivation

I'm hoping that the dependancy on mixins can be completely removed by replacing their usage with components. That would allow for the usage of `React.createClass` to be fully replaced with ES6 syntax & `extends component` / `extends pureComponent` and get us a step closer to React v16.

### Testing

To test these changes I've been doing the following:

- Go to a blog post
- Select 'Featured Image' > 'Set Featured Image' from the side-bar
  featured-image
- Upload or select an image from your media library
- follow through to the image editor for that image
  image-editor
- There's now a 'stack' of escapable components, `dialog-base` and `image-editor`
- Press esc once, the image editor should close.
- Press esc again, the media library (uses dialog-base) should now close too.
- Repeat this process but hold 'esc' down.
- Both components should close in succession.

cc'ing for continuation of #10268: @dmsnell, @gwwar, @gziolo, @retrofox, @ockham - FFTI though.

(I may still need to give https://github.com/Automattic/wp-calypso/pull/10268#issuecomment-275473195 and some other points some attention)

### CODE REVIEW

**Comment by gwwar** [comment]
@hoverduck any chance Flow Patrol could help test some of the cases here? We'll want to verify behavior <kbd> esc </kbd> behavior on:
- sidebar sites picker 
- image editor
- popovers
- all other dialog instances

**Comment by rachelmcr** [comment]
I tested the cases above (including every dialog instance I could find) and everything worked as expected compared to production.

There were a few cases where I expected esc to close something and it didn't. In these cases, the behavior is the same in production or it was a feature that hasn't launched in production yet. I'll note these cases for reference, even if we don't change the behavior right now:

- Editor: Toolbar dropdown menus (the ones with a down arrow) for Insert Content, Paragraph Styles, and Text Color
- Editor: Plain text mode explanatory dialog
- Editor: Post revisions (can't use esc to go back from post revision to post settings)
- Comments: Comment detail (after expanding a comment to see its detail, can't use esc to collapse it)
- Customizer: Can't use esc to close the customizer itself

Screenshots:

screen shot 2017-08-16 at 5 01 53 pm

screen shot 2017-08-16 at 5 02 41 pm

screen shot 2017-08-16 at 5 06 09 pm

screen shot 2017-08-16 at 5 47 20 pm

screen shot 2017-08-16 at 5 48 18 pm

**Comment by gwwar** [comment]
Thanks @rachelmcr great catches. I can write up an enhancement issue for Comments Management.

**Comment by spen** [comment]
@hoverduck 
Re: the flash of changing UI before the modal closes:

I've been doing some digging and this looks to be an existing bug. 
At first I thought that perhaps hitting <kbd>esc</kbd> was triggering 2 events, one switching the view and one closing the modal but it looks like it's a state issue.
When the modal closes the 'view' state is set back to `null` so that the next time it renders it doesn't maintain it's last place.
This happens when hitting cancel from the detail view (same flow as in the test plan but using the cancel button vs <kbd>esc</kbd> key).
I've made a note to circle back around to this at some point and will pop open a new issue for it. It seems like quite a confusing transaction :/

**Comment by spen** [comment]
Also thanks heaps @hoverduck & @rachelmcr for giving this a run through <3

> the behavior is the same in production or it was a feature that hasn't launched in production yet.

To be clear, these are issues seen in the app currently without this branch applied?
 I could pop open some issues to explore solving and ref them in this PR.

**Comment by rachelmcr** [comment]
>To be clear, these are issues seen in the app currently without this branch applied?

Yep, exactly. They aren't specific to this branch, so IMO they'd be nice to fix if possible but they aren't regressions here.

**Comment by gwwar** [comment]
> Re: the flash of changing UI before the modal closes:

@spen since this is an existing issue, as you suggested, let's write up an issue to document and not let it block the PR.

If you rebase I can take the branch for a last spin. Thanks to @hoverduck and @rachelmcr for testing too!

**Comment by spen** [comment]
Thanks @gwwar, #17418 covers the issue we see with the media modal - I'll get to rebasing tomorrow :)

File: client/components/close-on-escape/index.jsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		! isInput( event.target )
+	) {
+		const component = last( components );
+
+		component.onEscape();
+	}
+}
+
+function isInput( element ) {
+	return includes( [ 'INPUT', 'TEXTAREA' ], element.nodeName );
```
Comment (spen) [inline client/components/close-on-escape/index.jsx]: When testing this change (over the original `-1 !== [ 'INPUT', 'TEXTAREA' ].indexOf( element.nodeName );`), that the modal still closes when using one of the inputs in the `media-modal`s detail view.

I commented out the call to `.onEscape` in `onKeyDown` like so, yet still saw the modal close:

```js
function onKeydown( event ) {
	if (
		! isEmpty( components ) &&
		event.keyCode === ESC_KEY_CODE &&
		! isInput( event.target )
	) {
		const component = last( components );

		component.onEscape();
	}
}
```

I'll dig in to this a little more before merging this, testing the original mixin and looking for other listeners on the <kbd>esc</kbd> key.

Comment (spen) [inline client/components/close-on-escape/index.jsx]: Turns out that we're passing the _close method to the `react-modal` component as it's `onRequestClose` prop.
I'll adjust as necessary :)

**Comment by spen** [comment]
I've made some changes in the git history here that essentially remove `<CloseOnEscape />`s usage from `<DialogBase />` based on this train of thought: https://github.com/Automattic/wp-calypso/pull/17150#discussion_r134456693.
This renders the current test plan irrelevant as it no longer touches on `<CloseOnEscape />`.

I tested both removing `<CloseOnEscape />` and removing the `onRequestClose={ this._close }` prop from the `react-modal` instance inside `<DialogBase />` and the former seemed to make the most sense, at least for the time being.
The only issue that I see with leaving the closing of the modal up to `react-modal` is that we cannot check if the focused element is an input or not (as in the logic in the original `CloseOnEsc` mixin).

I'll write a new test plan a little later but would love feedback on the above decision in the mean-time if anybody has any :)

**Comment by gwwar** [comment]
@spen Could you elaborate on a bit on the bug you found? Basically we could still close the dialog using <kbd>esc</kbd> with the input focused?

**Comment by spen** [comment]
@gwwar sure thing!

I was testing a quick refactor of the `isInput` method and noticed, well, basically what you said - focusing on an input and clicking `esc` would still close the modal.

I did some digging there and as part of my debugging just removed `<CloseOnEscape />`... 
The modal still closed so I turned my attention to the Modal itself (an instance of `react-modal`), noticing that we pass `this._close` as the `onRequestClose` handler.

`react-modal`s <kbd>esc</kbd> handling doesn't account for inputs - that's why it was closing even with an input focused.

I see two options - 
1) we don't let `react-modal` handle the closing of the modal at all and let `<CloseOnEscape />` handle it. We would then need to handle clicks outside of the modal since `react-modal` handles this in the same breath as the <kbd>esc</kbd> handler (with no way to disable the <kbd>esc</kbd> handler).

2) We remove `<CloseOnEscape />` from `media-modal`. 
This is what I went with above but I'm finding other, subtle bugs with this approach. 
When opening the image editor from the media modal (following the test plan laid out above) and then closing it using <kbd>esc</kbd>, we can no longer just press <kbd>esc</kbd> again to close the media-modal - I assume it's no longer focused in the way `react-modal` expects.

Though I first thought that the second approach would be best, I'm now thinking the first could be so I hope to give that a go soon.

**Comment by spen** [comment]
Another alternative would be to add the ability to disable `react-modal` from closing on <kbd>esc</kbd> press [upstream](https://github.com/reactjs/react-modal/issues/481).
We could then allow react-modal to handle closing by clicking outside and handle the <kbd>esc</kbd> key ourselves.

**Comment by gwwar** [comment]
Unless, I'm misreading this sounds like an existing bug right? Since this is a good chunk of work already I'd be fine with opening another issue and addressing a fix for <kbd>esc</kbd> there.

Let's continue the conversation for the <kbd>esc</kbd> in another PR. Very quickly, your third option sounds the cleanest, but may take a while to land. Have we considered enhancing the props close method we pass to ReactModal? Another quick fix would be to maybe stop propagation of `esc` in our Modal wrapper.

**Comment by spen** [comment]
> Have we considered enhancing the props close method we pass to ReactModal?

It crossed my mind, but I've not explored it... I have a feeling it would make for some funky logic though - checking to see if it's a key press event or not, handling in one way if it is, another if it's not etc. Not as clean as I'd like but totally plausible.

It is an existing bug so happy to address separately, especially since it's so minor :)

**Comment by spen** [comment]
Issue opened for existing bugs @ #17534

**Comment by spen** [comment]
PR opened upstream to allow for the esc key to be disabled through a prop on `react-modal`: https://github.com/reactjs/react-modal/pull/487

File: client/components/close-on-escape/index.jsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import PropTypes from 'prop-types';
+import { Component } from 'react';
+
+const ESC_KEY_CODE = 27;
+let components = [];
+
+function onKeydown( event ) {
+	if (
+		! isEmpty( components ) &&
+		event.keyCode === ESC_KEY_CODE &&
```
Comment (marekhrabe) [inline client/components/close-on-escape/index.jsx]: Although `event.keyCode` is widely supported and may stay with us forever, [it's actually deprecated now](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode). React [does a great job of normalizing `event.key`](https://github.com/facebook/react/blob/master/src/renderers/dom/shared/utils/getEventKey.js) and I believe it's the preferred way to test for "special" keys. I think we should use that here and maybe later update our whole codebase for similar cases.

Comment (marekhrabe) [reply client/components/close-on-escape/index.jsx]: Oh sad, just noticed this is outside of React event system. In that case, we are probably better with keyCode. Ignore me

Comment (spen) [reply client/components/close-on-escape/index.jsx]: Good call out though - we could always put some logic around this if the need arises :)

**Comment by spen** [comment]
Wee update: The upstream PR has been merged (:tada:!) so once we're using an updated version we can handle our own esc key presses properly and iron out some of the little bugs mentioned :)

**Pull request merged**